src/main/scala/org/allenai/plugins/DockerBuildPlugin.scala

```diff
@@ -156,9 +164,59 @@ object DockerBuildPlugin extends AutoPlugin {
   }
 
   //////////////////////////////////////////////////////////////////////////////////////////////////
-  // Common settings used across tasks.
+  // Common settings / tasks / utilities used across tasks.
   //////////////////////////////////////////////////////////////////////////////////////////////////
 
+  /** Builds a docker image in the given directory. Before building, hash the contents, and check to
+    * see if the hash is different from what's in the given hash file. If the contents are
+    * unchanged, docker is not invoked. If the contents have checked, the old image is untagged, and
+    * the new image is built with the tags "latest" and a tag with the contents hash.
+    * @return the full image name, with the content hash tag included
+    */
+  def buildImageIfUpdated(
+    imageDir: File,
+    imageName: String,
+    hashFile: File,
+    logger: Logger
+  ): String = {
+    // Calculate the checksum of the contents of the main image.
+    val allFiles = PathFinder(imageDir).***.filter(_.isFile).get
+    val newHash = Utilities.hashFiles(allFiles, imageDir)
+
+    val oldHash = if (hashFile.exists) {
+      IO.read(hashFile)
+    } else {
+      ""
+    }
+
+    val imageNameWithLabel = imageName + ':' + newHash
+    if (newHash != oldHash) {
+      // Build a new docker image.
+      val buildCommand =
+        Seq("docker", "build", "-t", imageName, "-t", imageNameWithLabel, imageDir.toString)
```

Comment:
Does the naked image tag (`:latest` ) get pushed to the registry when pushing the SHA-tagged one, or do they have to be pushed separately?  I'm hoping `:latest` is a local-only tag.